File: Course_2-Data_Structures/Week-1/Excercise_Challenges/3_network_simulation/process_packages.cpp
#include <iostream>
#include <queue>
#include <vector>
#define PB push_back
#define MP make_pair
#define F first
#define S second
using namespace std;

int s;
queue<int> q;

int Processing(pair<int, int> p)
{
    while(!q.empty() && q.front() <= p.F)
        q.pop();

    if(q.size() < s){
        if(q.empty()){
            q.push(p.F + p.S);
            return p.F;
        }
        else{
            int st = max(p.F, q.back());
            q.push(st + p.S);
            return st;
        }
    }
    else{
        return -1;
    }
}

vector<int> ProcessRequests(vector<pair<int, int> > v)
{
    vector<int> response;
    for(int i = 0; i < v.size(); ++i){
        response.PB(Processing(v[i]));
    }

    return response;
}

int main(void)
{
    int n;
    cin >> s >> n;

    vector<pair<int, int> > v;
    for(int i = 0, start, process; i < n; ++i){
        cin >> start >> process;
        v.PB(MP(start, process));
    }

    vector<int> result = ProcessRequests(v);

    for(int i = 0; i < result.size(); ++i)
        cout << result[i] << endl;

    return 0;
}

File: Course_2-Data_Structures/Week-1/Excercise_Challenges/3_network_simulation/process_packages.py
size = 0
q = []

def process(v):
    while len(q) > 0 and q[0] <= v[0]:
        q.pop(0)

    if len(q) < size:
        if len(q) == 0:
            q.append(v[0] + v[1])
            return v[0]
        else:
            st = max(v[0], q[-1])
            q.append(st + v[1])
            return st
    else:
        return -1

def request_process(A):
    response = []
    for v in A:
        response.append(process(v))

    return response

if __name__ == "__main__":
    size, n = map(int, input().split())

    A = []
    for _ in range(n):
        s, p = map(int, input().split())
        A.append([s, p])

    response_list = request_process(A)

    for response in response_list:
        print(response)
